src/town_cmd.cpp

```diff
@@ -3132,8 +3135,88 @@ static void UpdateTownRating(Town *t)
 	SetWindowDirty(WC_TOWN_AUTHORITY, t->index);
 }
 
-static void UpdateTownGrowRate(Town *t)
+
+/**
+ * Updates town grow counter after growth rate change.
+ * Preserves relative house builting progress whenever it can.
+ * @param town The town to calculate grow counter for
+ * @param prev_growth_rate Town growth rate before it changed (one that was used with grow counter to be updated)
+ */
+static void UpdateTownGrowCounter(Town *t, uint16 prev_growth_rate)
+{
+	if (t->growth_rate == TOWN_GROWTH_RATE_NONE) return;
+	if (prev_growth_rate == TOWN_GROWTH_RATE_NONE) {
+		t->grow_counter = min(t->growth_rate, t->grow_counter);
+		return;
+	}
+	t->grow_counter = RoundDivSU((uint32)t->grow_counter * (t->growth_rate + 1), prev_growth_rate + 1);
+}
+
+/**
+ * Calculates amount of active stations in the range of town (HZB_TOWN_EDGE).
+ * @param town The town to calculate stations for
+ * @returns Amount of active stations
+ */
+static int CountActiveStations(Town *t)
+{
+	int n = 0;
+	const Station *st;
+	FOR_ALL_STATIONS(st) {
+		if (DistanceSquare(st->xy, t->xy) <= t->cache.squared_town_zone_radius[0]) {
+			if (st->time_since_load <= 20 || st->time_since_unload <= 20) {
+				n++;
+			}
+		}
+	}
+	return n;
+}
+
+/**
+ * Calculates town growth rate in normal conditions (custom growth rate not set).
+ * @param town The town to calculate growth rate for
+ * @returns Calculated growth rate
+ */
+static uint GetNormalGrowthRate(Town *t)
 {
+	static const uint16 _grow_count_values[2][6] = {
+		{ 120, 120, 120, 100,  80,  60 }, // Fund new buildings has been activated
+		{ 320, 420, 300, 220, 160, 100 }  // Normal values
+	};
+
+	int n = CountActiveStations(t);
+	uint16 m = _grow_count_values[t->fund_buildings_months != 0 ? 0 : 1][min(n, 5)];
```

Comment:
this line is fairly ugly, split it up a bit?

Reply:
Agree with @LordAro here, both array indexers should be calculated in separate variables with descriptive names. E.g. `fund_buildings_active` and `growth_rate_tier`. (Also swap the two inner arrays, it seems strange to me that the exceptional case comes first.)

Reply:
Well, I think having a clear logic is more important than pretty lines. And compared to the rest of codebase that line isn't even remotely complex. Like, sure, I can split it in 10 lines but it would only obfuscate the whole picture, we're just choosing one of predefined constants. And if someone needs to understand exactly which one he can still parse that line easily. As for line swapping it's completely irrelevant to this patch, if you think it's better that way you can make your own patch.